helpers/remote.py

```diff
@@ -345,15 +362,51 @@ def _connect(self):
             self._ssh.connect(
                 hostname=self.host, username=self.user, port=self.port, timeout=DEFAULT_TIMEOUT,
             )
+        except (paramiko.ssh_exception.SSHException, socket.error)as ssh_exc:
+            self._logger.error(
+                f"Failed to connect by SSH {self.host}:{self.port} by loading host keys from a system.",
+            )
+
+            if self._ssh_key:
+                self._connect_with_explicit_keys()
+
+            else:
+                self._logger.warning("SSH key was not provided, cannot try to connect with explicit keys.")
+                raise ssh_exc
+
         except Exception as conn_exc:
             self._logger.exception(f"Error connecting to {self.host} by SSH: {conn_exc}")
 
+    def _connect_with_explicit_keys(self):
+        """Open SSH connection to node with provided keys."""
+        if self._ssh_key:
```

Comment:
Is is duplicate `if` construction for `if` from `_connect_by_loading_keys_from_system` method. You will never get here if the `self._ssh_key` is False

Reply:
This method may be invoked separately and it is the main method to work with `_ssh_key`. I agree that there is duplication, I will remove the `if` from `_connect_by_loading_keys_from_system`.

Reply:
I remembered why I placed two `if` (short - extra validation):
1. the first `if` is needed to re-raise initial exception (to notify an user) in  `_connect_by_loading_keys_from_system` if `_ssh_key` was not provided - there is no sense to try to connect using a key in case when it was not provided.
2. the second `if`, as I already mentioned, the method may be invoked separately and `key_filename` is an optional arg for the `connect` method (from paramiko) and it may hide an error